lmdeploy/pytorch/paging/block_trie.py

```diff
@@ -0,0 +1,216 @@
+
+        while num_matched + block_size < seq.num_all_ids:
+            curr_tokens = seq.history_cache[num_matched:num_matched +
+                                            block_size]
+
+            key = hash(tuple(curr_tokens))
+            if key not in curr.children:
+                break
+
+            child = curr.children[key]
+            if not np.array_equal(curr_tokens, child.tokens):
+                break
+
+            __match_success(child)
+
+        if len(matched_blocks) > 0:
+            matched_blocks = np.array(matched_blocks)
+            self.allocator.update_access_time(matched_blocks)
+            self.allocator.add_ref_count(matched_blocks, 1)
+            seq.logical_blocks.append(matched_blocks)
+            seq.set_step(num_matched)
+
+        seq.logical_blocks.last_shared_node = curr
+
+    def allocate(self, seq: SchedulerSequence):
+        """allocate."""
+        if not self.enable:
+            return
+
+        block_size = self.block_size
+        logical_blocks = seq.logical_blocks
+        node: Node = getattr(logical_blocks, 'last_shared_node', None)
+        if node is None:
+            node = self.get_root(seq.adapter_name)
+            logical_blocks.last_shared_node = node
+
+        num_matched = node.num_matched
+        num_all_ids = seq.num_all_ids
+
+        if num_matched + block_size > num_all_ids:
+            return
+
+        if len(node.children) == 0 and node.parent is not None:
+            self.leaves.remove(node)
+
+        block_id = num_matched // block_size
+        blocks = []
+        free_blocks = []
+        while num_matched + block_size <= num_all_ids:
+            curr_tokens = seq.history_cache[num_matched:num_matched +
+                                            block_size]
+
+            block = logical_blocks[block_id]
+
+            hash_key = hash(tuple(curr_tokens))
+            parent = node
+            if hash_key in parent.children:
+                child = parent.children[hash_key]
+                if not np.array_equal(curr_tokens, child.tokens):
+                    break
+                node = child
+                free_blocks.append(block)
+                logical_blocks[block_id] = node.block
+            else:
+                node = Node(hash_key=hash_key,
+                            block=block,
+                            tokens=curr_tokens,
+                            num_matched=num_matched + block_size)
+                node.parent = parent
+            blocks.append(node.block)
+            num_matched += block_size
+            block_id += 1
+
+        logical_blocks.last_shared_node = node
+        if node.parent is not None and len(node.children) == 0:
+            # ignore root
+            self.leaves.add(node)
+        if len(blocks) > 0:
+            self.allocator.add_ref_count(np.array(blocks), 1)
+        if len(free_blocks) > 0:
+            self.allocator.free(np.array(free_blocks))
+
+    def evict(self, max_num_blocks: int):
```

Comment:
@grimoire Is there any starvation problem (https://github.com/InternLM/lmdeploy/pull/1450#issuecomment-2081421390) in Pytorch engine?

Reply:
No, I guess.
Running sequence would change it's status to Waiting if no enough blocks available. And sequence that require all blocks would be aborted https://github.com/grimoire/lmdeploy/blob/467c7972924c0d4faa226485f157a151debcf790/lmdeploy/pytorch/paging/scheduler.py#L252

That should be enough to prevent starvation.